src/main/java/com/metamx/common/FileUtils.java

```diff
@@ -0,0 +1,291 @@
+    private final Path path;
+    private volatile long childSize = -1l;
+    private volatile long lastSizeCheck = 0l;
+    private volatile long childModCount = 0l;
+    private final MessageDigest digest;
+    private final byte[] digestBytes;
+
+    public Collection<FileCopyResult> getChildren()
+    {
+      return ImmutableList.copyOf(children);
+    }
+
+    public MessageDigest getDigest()
+    {
+      return digest;
+    }
+
+    public Path getPath()
+    {
+      return path;
+    }
+
+    // Consumes otherDigest.digest()
+    public Boolean digestEquals(MessageDigest otherDigest){
+      if(this.digest == null || this.digestBytes == null){
+        return false;
+      }
+      if(!otherDigest.getAlgorithm().equals(digest.getAlgorithm())){
+        throw new IAE("Digest algorithm did not match. Expected [%s] got [%s]", digest.getAlgorithm(), otherDigest.getAlgorithm());
+      }
+      return Arrays.equals(digestBytes, otherDigest.digest());
+    }
+
+    // Only works for immutable children contents
+    public synchronized Long size()
+    {
+      final File file = path.toFile();
+      Long size = (file.exists() && !file.isDirectory()) ? file.length() : 0l;
+      if (childSize < 0l || childModCount > lastSizeCheck) {
+        Long childSize = 0l;
+        lastSizeCheck = childModCount;
+        for (FileCopyResult child : children) {
+          childSize += child.size();
+        }
+        this.childSize = childSize;
+      }
+      size += childSize;
+      return size;
+    }
+
+    // This consumes digest.digest()
+    public FileCopyResult(Path path, MessageDigest digest, FileCopyResult... children)
+    {
+      this.path = path;
+      this.children.addAll(Arrays.asList(children));
+      this.digest = digest;
+      this.digestBytes = digest.digest();
+    }
+    public FileCopyResult(Path path, Collection<FileCopyResult> children){
+      this.path = path;
+      this.children.addAll(children);
+      this.digest = null;
+      this.digestBytes = null;
+    }
+
+    public synchronized void addChild(FileCopyResult child)
+    {
+      this.children.add(child);
+      ++childModCount;
+    }
+
+    public static FileCopyResult buildResult(Path path, FileCopyResult... children) throws IOException{
+      return buildResult(path, "MD5", children);
+    }
+    public static FileCopyResult buildResult(Path path, String digestAlgorithm,  FileCopyResult... children) throws IOException{
+      final MessageDigest digest;
+      try {
+        digest = MessageDigest.getInstance(digestAlgorithm);
+      }
+      catch (NoSuchAlgorithmException e) {
+        throw new ISE(e, "Somehow I don't know how to do algorithm [%s]", digestAlgorithm);
+      }
+      try (final FileInputStream fileInputStream = new FileInputStream(path.toFile())) {
+        try (final DigestInputStream digestInputStream = new DigestInputStream(fileInputStream, digest)) {
+          final byte[] buffer = new byte[1 << 10];
```

Comment:
where does the size come from? can we declare it as a constant if it is constant?